### PR TITLE
Backport flavor extra spec

### PR DESCRIPTION
**Pull request by tj90241**
This commit backports the 'hw:vif_multiqueue_enabled'
flavor extra spec to Bionic/Ussuri.

Specifically, this PR:
https://review.opendev.org/c/openstack/nova/+/792356

is backported here. The way we have been applying patches
currently is to copy the full file out of Ussuri and apply the
necessary changes, which inflates the size of the diff/PR here.
In actuality, the only "net-new" code outside of the upstream
PR is the Chef changes here.

### CODE REVIEW

**Comment by tj90241** [comment]
Testing performed:
- New build
- Chef'ing existing build

If `hw:vif_mulitqueue_enabled` (flavor property) and `hw_vif_mulitqueue_enabled` (image property) are both specified AND disagree, THEN:
```shell
vagrant@r1n1:~$ openstack server create --min 1 --max 1 --network dev --image cirros --flavor dev-generic1.tiny test
Flavor dev-generic1.tiny has hw:vif_multiqueue_enabled extra spec explicitly set to True, conflicting with image cirros which has hw_vif_multiqueue_enabled explicitly set to False. (HTTP 400) (Request-ID: req-2274ed33-7456-4898-93be-5f2503597adb)
```

If either `hw:vif_mulitqueue_enabled` (flavor property) and `hw_vif_mulitqueue_enabled` (image property) is specified, then the VM is build in accordance to the property specified.

If neither `hw:vif_mulitqueue_enabled` (flavor property) and `hw_vif_mulitqueue_enabled` (image property) is specified, then multiqueue is not leveraged as before.

An administrator wishing to ensure all VMs have multiqueue enabled can control flavor properties such that all flavors always have `hw:vif_mulitqueue_enable=true`. In this case, it is not possible to launch a VM without multiqueue, even if some users have the capabilities to publish images.

**Comment by tj90241** [comment]
Also tested:
* Building a set of VMs without `hw:vif_multqueue_enabled=true` property set. Then live migrating (hot and cold), rebuilding, etc. Works fine, and the VM retains whatever multiqueue configuration it was originally built with.
* Resizing a VM that did not have multiqueue capabilities, but which is being resized to a flavor which does have mulitqueue. Also works as expected, and results in the target VM obtaining multiqueue (albeit while being rebooted in the process).

**Comment by comay** [comment]
Happy to review this when you're ready. Is this a straight backport with no behavior changes from Yoga?

**Comment by tj90241** [comment]
It is a nearly clean lift and shift, sans omitting some type-hinting from the patch (which only started being used in a later OpenStack release than Ussuri), and accounting for a variable name that changed in one function.

I was intending to hold off on making this ready for review until we have a way to assign multiqueue properties to flavors via automation, but can just make that effort a separate PR and keep this as more of an enablement PR of sorts.

**Comment by tj90241** [comment]
I can add the code to see if existing flavors don't yet have multiqueue and lob it in conditionally, but it makes a bit of the recipe and will just be more stuff to port to Chef. I'm thus proposing that we just make multiqueue the default for new flavors, and bolt on the property to existing clusters as needed during a deployment.

**Review by justinjpacheco — DISMISSED** [review]

**Comment by tj90241** [comment]
no changes, just rebased.

**Review by justinjpacheco — DISMISSED** [review]

**Review by psipika — DISMISSED** [review]

**Review by jingqgeng — APPROVED** [review]

**Review by comay — COMMENTED** [review]
I think an argument could be made to delivery all the patched files in both `nova-compute.rb` and `nova-head.rb` so that the files are consistent across all of the nodes even if all of them aren't actually imported/used but that can wait for another day.

(Perhaps we can move all of the "file delivery" into a separate file and then just restart the appropriate services in the individual recipes but again, that's another day.)

File: chef/cookbooks/bcpc/files/default/nova/vif.py [hunk 1]
```diff
@@ -274,18 +278,6 @@ def _get_virtio_mq_settings(self, image_meta, flavor):
...
-
-        if not isinstance(image_meta, objects.ImageMeta):
-            image_meta = objects.ImageMeta.from_dict(image_meta)
-
-        img_props = image_meta.properties
-
-        if img_props.get('hw_vif_multiqueue_enabled'):
-            return True
-
-        return False
```
[the remark this answers is not in the document]
Comment (tj90241) [reply chef/cookbooks/bcpc/files/default/nova/vif.py]: fixed

**Review by comay — APPROVED** [review]
Nice work, Tyler.

**Pull request merged**